### PR TITLE
add cache

### PR DESCRIPTION
**Pull request by jamesward**
Note that I wanted to preserve a way to use the locator statically so there is now a `WebJarVersionLocator.DEFAULT` for that.

### CODE REVIEW

File: src/main/java/org/webjars/WebJarVersionLocator.java [hunk 1]
```diff
@@ -17,70 +18,112 @@ public class WebJarVersionLocator {
     /**
      * The path to where webjar resources live.
      */
-    public static final String WEBJARS_PATH_PREFIX = "META-INF/resources/webjars";
+    public final String WEBJARS_PATH_PREFIX = "META-INF/resources/webjars";
```
Comment (dsyer) [inline src/main/java/org/webjars/WebJarVersionLocator.java]: These could (should) still be static, right? Not sure what the intent was here.

File: src/main/java/org/webjars/WebJarVersionLocator.java [hunk 2]
```diff
@@ -17,70 +18,112 @@ public class WebJarVersionLocator {
...
+    }
+
+    public WebJarVersionLocator(WebJarCache cache) {
+        this.cache = cache;
+    }
 
-        if (LOADER.getResource(fullPath) != null) {
-            return fullPath;
+    public static class DEFAULT {
+        private static final WebJarVersionLocator webJarVersionLocator = new WebJarVersionLocator();
```
Comment (dsyer) [inline src/main/java/org/webjars/WebJarVersionLocator.java]: Any reason why DEFAULT is a static class not a field? Just curious. The API for users is the same but I don’t see this idiom anywhere else.

**Review by dsyer — COMMENTED** [review]
Kind of OK by me. See comments.

File: src/main/java/org/webjars/WebJarVersionLocator.java (hunk 1 above)
Comment (jamesward) [reply src/main/java/org/webjars/WebJarVersionLocator.java]: I was thinking we may want the opportunity to expose an override (constructor param) at some point in that instance. But maybe that isn't necessary.

File: src/main/java/org/webjars/WebJarVersionLocator.java (hunk 2 above)
Comment (jamesward) [reply src/main/java/org/webjars/WebJarVersionLocator.java]: Yeah, I think you are right that it could just be a field.  I'll give that a try.

**Comment by jamesward** [comment]
Thanks @dsyer for the review!

File: src/main/java/org/webjars/WebJarVersionLocator.java (hunk 2 above)
Comment (jamesward) [reply src/main/java/org/webjars/WebJarVersionLocator.java]: Ok, this change is now in.

File: src/main/java/org/webjars/WebJarVersionLocator.java (hunk 1 above)
Comment (sdeleuze) [reply src/main/java/org/webjars/WebJarVersionLocator.java]: I would be in favor of keeping these static as well. The uppercase suggest by convention these values are constants  so if we change them that could be confusing. You can maybe introduce non static field later when needed, here that's more to communicate the defaults.

**Comment by sdeleuze** [comment]
I can live with it, but I am not a big fan of the `DEFAULT` static instance:
 - I don't think `WebJarAssetLocator ` from `webjars-locator-core` had one
 - I expect a lot of users just missing it and using `new WebJarVersionLocator()` leading to 2 instances created with 1 not used
 - If we are afraid users doing `new WebJarVersionLocator()` for each request, we could maybe just remove the static field and document this properly in the README (like `webjars-locator-core` does) and in the Javadoc.

Otherwise looks to to me.

**Comment by jamesward** [comment]
Thanks @sdeleuze. Yeah, my concerns were that'd it is too easy to create a new cache on each usage.  I'm not sure if there is a better design which has a static default and a way to override it.  Any typical patterns for this?

**Comment by sdeleuze** [comment]
I don't think there is a good "self-contained" way to do that, by design. But I am not sure the situation is different than `WebJarAssetLocator` which had heavy resource scanning at initialization. I would even say that the cost of a misuse is lower now.

My recommandation would be to only provide the constructor, without the static field, and document clearly on both Javadoc and README that:
 -  `WebJarVersionLocator` is thread-safe
 - That a single instance (singleton) is expected to be created, either via a static field (a code sample showing a `WebJarVersionLocator` static field on user class can be provided) or via other mechanism like dependency injection. 
 
This is also what is done for very popular dependency of the ecosystem like Jackson `ObjectMapper`.

**Comment by jamesward** [comment]
I guess in Java that is the best we can do. :) I'll get those changes done today.

**Comment by jamesward** [comment]
Changes are in.  LMKWYT.  Thanks!

**Comment by sdeleuze** [comment]
Awesome, looks good to me for the singleton removal.

The only remaining ask would be to make `PROPERTIES_ROOT`, `NPM`, `PLAIN` and `POM_PROPERTIES` `static` like what you did for `WEBJARS_PATH_PREFIX` (I think this was the spirit of @dsyer review as he mentioned "**These** could (should) still be static, right?" even if I was also confused by the fact the feedback in the code review was just on `WEBJARS_PATH_PREFIX`).

File: src/main/java/org/webjars/WebJarVersionLocator.java [hunk 3]
```diff
@@ -19,68 +20,104 @@ public class WebJarVersionLocator {
      */
     public static final String WEBJARS_PATH_PREFIX = "META-INF/resources/webjars";
 
-    private static final String PROPERTIES_ROOT = "META-INF/maven/";
-    private static final String NPM = "org.webjars.npm/";
-    private static final String PLAIN = "org.webjars/";
-    private static final String POM_PROPERTIES = "/pom.properties";
+    private final String PROPERTIES_ROOT = "META-INF/maven/";
```
Comment (jamesward) [inline src/main/java/org/webjars/WebJarVersionLocator.java]:
```suggestion
    private static final String PROPERTIES_ROOT = "META-INF/maven/";
```

File: src/main/java/org/webjars/WebJarVersionLocator.java [hunk 4]
```diff
@@ -20,67 +21,103 @@ public class WebJarVersionLocator {
     public static final String WEBJARS_PATH_PREFIX = "META-INF/resources/webjars";
 
     private static final String PROPERTIES_ROOT = "META-INF/maven/";
-    private static final String NPM = "org.webjars.npm/";
-    private static final String PLAIN = "org.webjars/";
-    private static final String POM_PROPERTIES = "/pom.properties";
+    private final String NPM = "org.webjars.npm/";
```
Comment (jamesward) [inline src/main/java/org/webjars/WebJarVersionLocator.java]:
```suggestion
    private static final String NPM = "org.webjars.npm/";
```

File: src/main/java/org/webjars/WebJarVersionLocator.java [hunk 5]
```diff
@@ -21,66 +22,102 @@ public class WebJarVersionLocator {
 
     private static final String PROPERTIES_ROOT = "META-INF/maven/";
     private static final String NPM = "org.webjars.npm/";
-    private static final String PLAIN = "org.webjars/";
-    private static final String POM_PROPERTIES = "/pom.properties";
+    private final String PLAIN = "org.webjars/";
```
Comment (jamesward) [inline src/main/java/org/webjars/WebJarVersionLocator.java]:
```suggestion
    private static final String PLAIN = "org.webjars/";
```

File: src/main/java/org/webjars/WebJarVersionLocator.java [hunk 6]
```diff
@@ -22,65 +23,101 @@ public class WebJarVersionLocator {
     private static final String PROPERTIES_ROOT = "META-INF/maven/";
     private static final String NPM = "org.webjars.npm/";
     private static final String PLAIN = "org.webjars/";
-    private static final String POM_PROPERTIES = "/pom.properties";
+    private final String POM_PROPERTIES = "/pom.properties";
```
Comment (jamesward) [inline src/main/java/org/webjars/WebJarVersionLocator.java]:
```suggestion
    private static final String POM_PROPERTIES = "/pom.properties";
```

**Comment by jamesward** [comment]
Cool. I've changed that. And what about:
```
private final ClassLoader LOADER = WebJarVersionLocator.class.getClassLoader();
```

Should it be `static` as well?

**Comment by sdeleuze** [comment]
Indeed it can be static as well.

File: src/main/java/org/webjars/WebJarVersionLocator.java [hunk 7]
```diff
@@ -24,63 +25,99 @@ public class WebJarVersionLocator {
     private static final String PLAIN = "org.webjars/";
     private static final String POM_PROPERTIES = "/pom.properties";
 
-    private static final ClassLoader LOADER = WebJarVersionLocator.class.getClassLoader();
+    private final ClassLoader LOADER = WebJarVersionLocator.class.getClassLoader();
```
Comment (jamesward) [inline src/main/java/org/webjars/WebJarVersionLocator.java]:
```suggestion
    private static final ClassLoader LOADER = WebJarVersionLocator.class.getClassLoader();
```

**Comment by vpavic** [comment]
@jamesward do you really anticipate that someone would want to use a custom cache implementation with `WebJarVersionLocator`?

It seems to me that caching can easily be an internal implementation detail, which then removes the need for `WebJarCache` abstraction and additionally allows your cache interactions to be atomic (by leveraging `ConcurrentHashMap#computeIfAbsent`).

**Comment by jamesward** [comment]
Good point.  If the Spring's usage doesn't foresee plugging in a custom cache then we could make it internal.  Question on `ConcurrentHashMap#computeIfAbsent` - I imagine that a common situation using the locator would be that multiple requests for the same WebJar will come in at the same time (due to parallel loading from the browser). Would `computeIfAbsent` essentially do a lock so that only 1 of the concurrent requests would do the actual lookup?
cc @sdeleuze

**Comment by sdeleuze** [comment]
There is indeed likely something to refine here.

>  Would computeIfAbsent essentially do a lock so that only 1 of the concurrent requests would do the actual lookup?
For the initial request yes, see https://stackoverflow.com/a/10589196/1092077 for more details.

I created a dedicated new issue, let's maybe continue the discussion there: https://github.com/webjars/webjars-locator-lite/issues/8

**Pull request merged**